### ISSUE TITLE
When v3 will be released?

### ISSUE THREAD

**Issue by vimmerru** [mikro-orm/mikro-orm#169]
Hi @B4nan do you have any formal date for v3?

**Comment by B4nan** [comment]
I was hoping to get it out till the end of September, but then @lookfirst jumped in and proposed a lot of improvements that I wanted to incorporate as well (thanks for that!), so it will be a bit delayed probably. Hopefully it will come during first half of October. 

I still have some things to implement, especially #124 and #121 as those will be breaking changes. Both are already prototyped somewhere in my local branches, but I will need few more days to finalize. 

Generally it should be already usable, I am tagging it as alpha just because its not feature complete (those 2 issues I mentioned above). I am using the latest alpha myself in production (nestjs + postgres), so far no problems. 

I just deployed the v3 docs to https://mikro-orm.github.io/docs-v3/index, so you don't have to browse thru github before it gets merged. (one gotcha, not sure why, but the index is not working properly so this throws 404: https://mikro-orm.github.io/docs-v3/). 

Once I will finish with all features and write a migration guide, I will merge it to master and release RC. Will keep it in RC till I have article ready about the release, which will also take some time (a week or so).

**Comment by lookfirst** [comment]
@vimmerru Use the 'next' npm version and give him feedback. It is pretty stable (although I keep finding weirdo edge case issues). That said it is a vast improvement over v2 and definitely the best TS ORM solution on the market.

**Comment by B4nan** [comment]
> although I keep finding weirdo edge case issues

Yeah but pretty much all of them are not really about v3 :] I think you would have even more troubles with v2 as the validation/error handling has improved a lot in v3.

**Comment by lookfirst** [comment]
... which is why when I found this project, I read the docs for v2, looked for the 'next' branch and installed that immediately. ;-)

**Comment by vimmerru** [comment]
ok, i will try )

**Comment by j** [comment]
@B4nan I'm an old timer who used to be super fond of all things doctrine so I was happy to see a similar type of ODM in the node space.  I first checked out MikroORM when it was mongo-only which was the best part about it.. Now I'm seeing SQL in the mix and it kinda makes me a little sad.  I'm afraid to choose an ORM that have many drivers since it can bog down real problems or features for a specific use-case.  With this rant above, is there any reason why you didn't do what Doctrine did and maybe turn this into a mono-repository that has a core logic and interfaces so that each actual driver (MongoDB vs SQL) can have it's separation.

Side Note: Since transactions support is in latest mongo versions, it's sad to see that it's not part of MikroORM.

**Comment by lookfirst** [comment]
See, I'm the opposite ... I can't imagine using mongo over postgres.... not trying to start a flame war, which is why I don't state 'why'... just giving a different perspective.

**Comment by j** [comment]
@lookfirst I'm not saying any one is better, I'm saying they are different so separation is better.  Documentation says it's inspired by DoctrineORM and I see that.  But there's DoctrineORM and DoctrineODM (for mongo) because they are *different* and they both use https://github.com/doctrine/common.

Would be cool to see similar here.  `mikro-common`, `mikro-orm` (sql) and `mikro-odm` (mongo).  In the mongo world, I like to just use straight mongo queries as do a lot of other people, whereas, ORMs for SQL with a query builder is nice to have.  But mainly, I was looking at the documentation and all the sections are SQL based now and I don't see anything MongoDB.

This project just looks to be that it used to be MongoDB focused, then the author switched over to using SQL and left MongoDB in the past.  Which is why I'd never use TypeORM's mongo driver... it's so terrible.

And regarding flame-wars... if you use a database and it makes a company millions of $, that's not a bad db. ;) . Use what pays bills, employs, and makes people happy.

**Comment by B4nan** [comment]
> is there any reason why you didn't do what Doctrine did and maybe turn this into a mono-repository

I would like to do that, but not a priority for now. From user's perspective not much will change, so I am rather focusing on features than refactoring. I would like to start with having one package per driver, making some common package might come as well, but currently most of the ORM's code is for all drivers. 

> Now I'm seeing SQL in the mix and it kinda makes me a little sad. I'm afraid to choose an ORM that have many drivers since it can bog down real problems or features for a specific use-case. 

Sorry to hear that, do you have something specific on your mind? Currently (apart from schema and entity generator) pretty much everything works same in all drivers (only difference is that mongo driver does not support searching by relations as there are no joins in mongo and $lookup is not very usable for this as it can be super slow when used like sql join). 

> Side Note: Since transactions support is in latest mongo versions, is that going to be supported?

Yes, planning to do that rather soon, not sure if before v3 is released, as I don't wanna postpone the release for too long. 

> I was looking at the documentation and all the sections are SQL based now

Well most of the docs are for all drivers, there might be sql only examples somewhere, but that does not mean it is not working with mongo too. Again, do you have some concrete links? Will be happy to improve that so it's not confusing. 

> This project just looks to be that it used to be MongoDB focused, then the author switched over to using SQL and left MongoDB in the past. 

Yeah that is kinda what happened, I was using mongo on all my projects, which is no longer true. But I wouldn't call it that I left mongo in the past. There is still 100% test coverage for all drivers including mongo one. I am open for feature requests there as well. 

> Which is why I'd never use TypeORM's mongo driver... it's so terrible.

I had the very same experience, which pushed me even more into development of this.

**Comment by j** [comment]
Also @B4nan, great work.  It's definitely the best ORM I've seen in the JS space.  I haven't looked too deep into it.  The `ts-morph` thing seems genius.

In all honesty, we have about $10k in credits for MongoDB right now so we've been just using that, but it's been a while since we've used SQL for anything other than BI.  And we're in the midst of refactoring our e-commerce platform so Postgres might come into discussion. :P

**Comment by lookfirst** [comment]
> Use what works, pays bills, employs, and makes people happy.

Of course! That's why I pick postgres over mongo. =) I did build a company that grossed $80m in our first year. We tried mongo and it fell over on us and postgres saved the day. Note that a very good friend of mine worked for mongo corp from very early on, so it wasn't like we didn't have inside expert advice on it.

I hear what you're saying though about splitting it up, but I totally agree with @B4nan ...  not until needed. If there is something you're missing, then suggest it as a feature or bug request.

Agreed it is the best ORM for JS/TS!
